### ISSUE TITLE
Refactor and clean code / New API

### ISSUE THREAD

**Issue by buschtoens** [buschtoens/dhcp#2]
Follow Up for #1.

Still have to think about a good API, that is easy to use and clean.

**Comment by maxired** [comment]
Hi,

I would like to get more information about your vision of the futur of the project.
From my point of view, it is very cool to implement a DHCP server with specific rules for IP assignation for example, depending of the node MAC, names, ....

In order to be more flexible I think that a middleware approach, commonly seen in JavaScript project such as in 'connect' or 'express' , would be a great solution.
After this I would like to create some middleware that could be easily reusable.

My reflection came to this point after that the example I am currently developping, [DNS-DHCP-server.js](https://github.com/maxired/dhcp/blob/master/examples/DNS-DHCP-server.js) , became too monolithic and specific.

**Comment by buschtoens** [comment]
Yep, you're right. Stuffing everything into one file sucks. I really like the middleware approach.
Only problem is, that DHCP isn't stateless as HTTP.

I suggest, that the server itself parses the request and recompiles the answer. The middleware, in order of registration, gets called for each request like this.

``` js
var req = {
       // basically the same stuff as below, just parsed from the actual request
     }
  , res = {
      header: {
        op: 2
      , htype: 1
      , hlen: 6
      , hops: 0
      , xid: req.header.xid
      , secs: ~
      , flags: ~
      , ciaddr: req.header.ciaddr
      , yiaddr: ~
      , siaddr: ~
      , giaddr: ~
      , chaddr: req.header.chaddr
      }
    , options: {
        // all options requested via 55 "Parameter Request List", but set to `null`
      }
    };
middleware(req, res, session, stop);
```

Calling `stop()` would make the server immediately send back the response. not going through further middleware that were registered later on. Pretty much the opposite of `next()`.
`session` should be an object that persistently saves data associated with one client. Client tracking should be done via MAC address (`chaddr`) as it should be included in every request.

Middleware will mainly set fields in `res.options`. I think the `session` object leaves some space for improvements. We'll find that out during development, when we see what's necessary or could be solved more elegantly.

What are your thoughts on this?

**Comment by maxired** [comment]
This seems pretty reasonable to me. Nevertheless, there are some points on which I want your opinion. 

With this approach, we can easily send only one message, after we received one. I don't have any use case in mind, but I am pretty sure this could be convenient to be able to send severals messages.  In express, we manually call res.send()  which is a good thing. 
Moreover,  if stop could be usefull in some cases,  a next seems to be mandatory if we want to make asynchronous call in a middeware. 

Concerning the session object, this is an important topic to work on. We should be able to decide of severals storage strategy, such as in memory or in a redis database. 

Maybe this would make stuff more complex, but I think that a middeware might need to know if an option it has setted has been overriden by a following middeware. This could be done with another callback mechanism,  or maybe with an event.

**Comment by buschtoens** [comment]
#### Multiple Messages

In DHCP, the client initialises a conversation with a `DHCPDISCOVER`, if it has no internal IP preconfiguration, or `DHCPREQUEST`, if wants to claim an internally specified IP and not be given one by the server via `DHCPOFFER`.
DHCP-Servers in the network then send `DHCPACK` back. The client only accepts the first one and finishes the initialization.

The normally nothing happens, until the client gets shutdown. On a graceful shutdown the client would send a `DHCPRELEASE` releasing the IP address.

Ther _should_ be no reason or possibility by design to subsequently send multiple messages to one client. This would violate DHCP's core concept. As we're currently aiming for a server which communicates with standard clients, we should obey to protocol. If you, at any time, can come up with a use case, I'd say we start a new branch, testing everything out and then - maybe - merge it into the master branch or release a new npm package.

However, `dhcp.broadcast(header, options)` is a neat idea. See #4.

#### `stop()` / `next()` & middleware interaction

You're right. I was just lazy thinking. We need to go for the `next()` approach. The only thing, that nags me about middleware is, that they don't really interact with each other and, more importantly, can't react to each other.
I think this poses a major problem, as the middlewares most likely will overwrite each other in production use. We need some control for this.
Just setting a field in `res` seemed easy, but doesn't suffice. What do you think about this:

``` js
res.options.set(DHCP.code.vendor.DOMAIN_NAME_SERVER, ["8.8.8.8", "8.8.4.4"], DHCP.priority.LOWEST);

// later on, this overwrites the first as DHCP.priority.LOWEST < DHCP.priority.HIGH
res.options.set(DHCP.code.vendor.DOMAIN_NAME_SERVER, ["208.67.222.222", "208.67.220.220"], DHCP.priority.HIGH);
```

The same concept shown here goes for `res.header`. This way, we could call all middlewares in parallel, not having to wait for each `next()`. We wait until all `next()`s were called and then send the response. Somewhat like [`async.parallel()`](https://github.com/caolan/async#parallel).

Nevertheless, middlewares should still be informed, when one of their settings is being overwritten. Maybe they could `return` a function that is called each time a setting set by the middleware is overwritten.
Glueing everything together we would end up with this code.

``` js
var dhcp = new DHCP;

// configure the server, express-like, optionally wrapped with dhcp.configure(), see http://expressjs.com/api.html#app.configure
dhcp.set("name", "value");
dhcp.enable("bool true");
dhcp.disable("bool false");

// wire up the middleware, pretty much like in express
dhcp.use(function(req, res, session, stop) {
  res.set(DHCP.code.vendor.DOMAIN_NAME_SERVER, ["8.8.8.8", "8.8.4.4"], DHCP.priority.LOWEST);

  return function(field, oldValue, newValue, overwrite, oldPriority, newPriority) {
      console.log(field, oldValue, newValue); // DHCP.code.vendor.DOMAIN_NAME_SERVER, ["8.8.8.8", "8.8.4.4"], somethingDifferent
      overwrite(); // Do nothing about it, same as calling `next()`
      overwrite(oldValue, DHCP.priority.NORMAL); // overwrite the newValue with the oldValue, if priority is high enough
    };
});
```

**Comment by maxired** [comment]
## stop() / next() & middleware interaction

To be honest, I am not a big fan of this approach.
If you call your middleware in parallels : 
- You can't use `stop`, because you don't know what's already  append or not.
- What happen if two middleware set a field with the same priority. tricky question since the order is not always the same.
- Setting are not independent from each other. for example `sname` and `file` are use together by the client. This make me think that the callback for each overidden value is maybe not the good choice.

=> all this reasons make me think that a solution with each callbacks called sequentially in the order of initialization would be better.
